### PR TITLE
Viewer for PV and PVS

### PR DESCRIPTION
**Pull request by ganchandrasekaran**

### CODE REVIEW

**Comment by ganchandrasekaran** [comment]
fix https://github.com/nephio-project/nephio/issues/184

**Comment by ganchandrasekaran** [comment]
fix https://github.com/nephio-project/nephio/issues/120

**Comment by ChristopherFry** [comment]
Left a few comments for the PackageVariant viewer on https://github.com/nephio-project/nephio/issues/119#issuecomment-1548157317. The comments should be applicable for the PackageVariantSet viewer as well.

File: plugins/cad/src/types/PackageVariantSet.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    repositorySelector: PackageVarientSetRepositorieSelector;
+    objectSelector: string;
+  };
+
+  export type PackageVarientSetRepositories = {
+    name: string;
+    packageNames?: string[];
+  }
+
+  export type PackageVarientSetRepositorieSelector = {
```
Comment (ChristopherFry) [inline plugins/cad/src/types/PackageVariantSet.ts:51]: nit: `RepositorySelector` versus `RepositorieSelector`

File: plugins/cad/src/components/ResourceViewerDialog/components/FirstClassViewers/StructuredMetadata/resources/packageVariant.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    variantAnnotations: getValue(packageVariant.spec.annotations),
+    upstream: packageVariant.spec.upstream
+      ? `${packageVariant.spec.upstream.repo}/${packageVariant.spec.upstream.package}@${packageVariant.spec.upstream?.revision 
+        ? packageVariant.spec.upstream?.revision
+        : '0'}`
+      : '',
+    downstream: packageVariant.spec.downstream
+      ? `${packageVariant.spec.downstream.repo}/${packageVariant.spec.downstream.package}@${packageVariant.spec.downstream?.revision 
+        ? packageVariant.spec.downstream?.revision
+        : '0'}`
```
Comment (ChristopherFry) [inline plugins/cad/src/components/ResourceViewerDialog/components/FirstClassViewers/StructuredMetadata/resources/packageVariant.ts:79]: The downstream package does not have a revision, so we only need `{repo}/{package}` here (revision can be removed).

https://github.com/GoogleContainerTools/kpt/blob/main/porch/controllers/packagevariants/api/v1alpha1/packagevariant_types.go#L76

File: plugins/cad/src/components/ResourceViewerDialog/components/FirstClassViewers/StructuredMetadata/resources/packageVariant.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  const contextdata: Metadata = {};
+  setContextdata(contextdata, packageVariant.spec.packageContext, "packageContext");
+
+  const customMetadata: Metadata = {
+    variantLabels: getValue(packageVariant.spec.labels),
+    variantAnnotations: getValue(packageVariant.spec.annotations),
+    upstream: packageVariant.spec.upstream
+      ? `${packageVariant.spec.upstream.repo}/${packageVariant.spec.upstream.package}@${packageVariant.spec.upstream?.revision 
+        ? packageVariant.spec.upstream?.revision
+        : '0'}`
```
Comment (ChristopherFry) [inline plugins/cad/src/components/ResourceViewerDialog/components/FirstClassViewers/StructuredMetadata/resources/packageVariant.ts:74]: I would avoid defaulting revision if one is not specified.

File: plugins/cad/src/components/ResourceViewerDialog/components/FirstClassViewers/StructuredMetadata/resources/packageVariant.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+export const getPackageVariantStructuredMetadata = (
+  resource: KubernetesResource,
+): Metadata => {
+  const packageVariant = resource as PackageVariant;
+  
+  const contextdata: Metadata = {};
+  setContextdata(contextdata, packageVariant.spec.packageContext, "packageContext");
+
+  const customMetadata: Metadata = {
```
Comment (ChristopherFry) [inline plugins/cad/src/components/ResourceViewerDialog/components/FirstClassViewers/StructuredMetadata/resources/packageVariant.ts:68]: Looks like `adoptionPolicy` and `deletionPolicy` are both missing from the custom metadata.

File: plugins/cad/src/components/ResourceViewerDialog/components/FirstClassViewerSelector.tsx [hunk 5]
```diff
@@ -98,6 +100,13 @@ const getCustomMetadataFn = (
     case 'v1/ConfigMap':
       return getConfigMapStructuredMetadata;
 
+
+    case 'config.porch.kpt.dev/v1alpha1/PackageVariant':
+      return getPackageVariantStructuredMetadata;
+
+    case 'config.porch.kpt.dev/v1alpha2/PackageVariantSet':
+      return getPackageVariantSetStructuredMetadata;
+
```
Comment (ChristopherFry) [inline plugins/cad/src/components/ResourceViewerDialog/components/FirstClassViewerSelector.tsx:109]: nit: let's move this to above  `case 'fn.kpt.dev/v1alpha1/ApplyReplacements'` to keep this list in alphabetical order with the group version kind.

**Review by ChristopherFry — APPROVED** [review]
Left a few comments but looks good overall.

Also, before submitting, please run `yarn prettier:write` locally. This ensures all the code is formatted consistently within the project.

**Pull request merged**